iothub/device/src/Transport/HttpClientHelper.cs

```diff
@@ -538,6 +541,20 @@ public void Dispose()
                     _httpClientHandler = null;
                 }
 
+                // The associated TokenRefresher should be disposed by the http client helper only when the http client
+                // is the primary transport handler.
+                // For eg. we create HttpTransportHandler instances for file upload operations even though the client might be
+                // initialized via MQTT/ AMQP. In those scenarios, since the shared TokenRefresher resource would be primarily used by the
+                // corresponding transport layers (MQTT/ AMQP), the diposal should be delegated to them and it should not be disposed here.
+                // The only scenario where the TokenRefresher should be disposed here is when the client has been initialized using HTTP.
+                if (_isClientPrimaryTransportHandler
+                    && _authenticationHeaderProvider is IotHubConnectionString iotHubConnectionString
+                    && iotHubConnectionString.TokenRefresher != null
+                    && iotHubConnectionString.TokenRefresher.InstanceCreatedBySdk)
+                {
+                    iotHubConnectionString.TokenRefresher.Dispose();
```

Comment:
if applicable, dispose the token refresher when the http client helper is disposed